src/server.cpp

```diff
@@ -1566,13 +1566,41 @@ void Server::SendShowFormspecMessage(session_t peer_id, const std::string &forms
                                      const std::string &formname)
 {
 	NetworkPacket pkt(TOCLIENT_SHOW_FORMSPEC, 0 , peer_id);
+	std::string client_name;
+
+	auto peer_state_iterator = m_formspec_state_data.find(peer_id);
 	if (formspec.empty()){
 		//the client should close the formspec
-		pkt.putLongString("");
+		if (peer_state_iterator != m_formspec_state_data.end()) {
+			PeerFormspecState peer_state = peer_state_iterator->second;
+			if (!formname.empty()) {
+				if (peer_state.formspec_name != formname) {
+					errorstream << "attempting to close formspec '" << formname
+							<< "', but a different formspec is open '"
+							<< peer_state.formspec_name << "'" << std::endl;
+					return;
+				}
+				client_name = peer_state.getClientSideName();
+			}
+			m_formspec_state_data.erase(peer_state_iterator);
+			pkt.putLongString("");
+		} else {
+			errorstream << "attempting to close formspec '" << formname
```

Comment:
Not `errorstream` here either. If an error or warning should be shown, then throw it with more information from the calling Lua mod.